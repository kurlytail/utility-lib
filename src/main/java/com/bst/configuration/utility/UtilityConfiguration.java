package com.bst.configuration.utility;

import org.springframework.context.annotation.ComponentScan;
import org.springframework.context.annotation.Configuration;

@Configuration
@ComponentScan("com.bst.utility")
public class UtilityConfiguration {

}
