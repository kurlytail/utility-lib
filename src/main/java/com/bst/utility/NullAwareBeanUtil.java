package com.bst.utility;

import java.util.HashSet;
import java.util.List;
import java.util.Set;

import org.springframework.beans.BeanUtils;
import org.springframework.beans.BeanWrapper;
import org.springframework.beans.BeanWrapperImpl;

public class NullAwareBeanUtil {
	// then use Spring BeanUtils to copy and ignore null
	public static void copyProperties(final Object src, final Object target) {
		BeanUtils.copyProperties(src, target, NullAwareBeanUtil.getNullPropertyNames(src));
	}

	private static String[] getNullPropertyNames(final Object source) {
		final BeanWrapper src = new BeanWrapperImpl(source);
		final java.beans.PropertyDescriptor[] pds = src.getPropertyDescriptors();

		final Set<String> emptyNames = new HashSet<>();
		for (final java.beans.PropertyDescriptor pd : pds) {
			final Object srcValue = src.getPropertyValue(pd.getName());
			if ((srcValue == null) || (srcValue instanceof Set<?>) || (srcValue instanceof List<?>)) {
				emptyNames.add(pd.getName());
			}
		}
		final String[] result = new String[emptyNames.size()];
		return emptyNames.toArray(result);
	}
}