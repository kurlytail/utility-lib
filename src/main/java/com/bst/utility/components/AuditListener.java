package com.bst.utility.components;

import javax.persistence.PostLoad;
import javax.persistence.PostPersist;
import javax.persistence.PostRemove;
import javax.persistence.PostUpdate;
import javax.persistence.PrePersist;
import javax.persistence.PreRemove;
import javax.persistence.PreUpdate;

import org.springframework.beans.factory.annotation.Autowired;

public class AuditListener {

	@Autowired
	private AuditService auditService;

	private void checkAuditService() throws Exception {
		if (this.auditService == null) {
			AuditService.autowire(this);
			if (this.auditService == null) {
				throw new Exception("Audit service not initialized");
			}
		}
	}

	@PostLoad
	public void postLoad(final Object object) throws Exception {
		this.checkAuditService();
		this.auditService.postLoad(object);
	}

	@PostPersist
	public void postPersist(final Object object) throws Exception {
		this.checkAuditService();
		this.auditService.postPersist(object);
	}

	@PostRemove
	public void postRemove(final Object object) throws Exception {
		this.checkAuditService();
		this.auditService.postRemove(object);
	}

	@PostUpdate
	public void postUpdate(final Object object) throws Exception {
		this.checkAuditService();
		this.auditService.postUpdate(object);
	}

	@PrePersist
	public void prePersist(final Object object) throws Exception {
		this.checkAuditService();
		this.auditService.prePersist(object);
	}

	@PreRemove
	public void preRemove(final Object object) throws Exception {
		this.checkAuditService();
		this.auditService.preRemove(object);
	}

	@PreUpdate
	public void preUpdate(final Object object) throws Exception {
		this.checkAuditService();
		this.auditService.preUpdate(object);
	}

}
