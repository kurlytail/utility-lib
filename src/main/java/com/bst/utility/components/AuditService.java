package com.bst.utility.components;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.HashMap;
import java.util.Map;

import javax.persistence.PostLoad;
import javax.persistence.PostPersist;
import javax.persistence.PostRemove;
import javax.persistence.PostUpdate;
import javax.persistence.PrePersist;
import javax.persistence.PreRemove;
import javax.persistence.PreUpdate;

import org.springframework.beans.BeansException;
import org.springframework.context.ApplicationContext;
import org.springframework.context.ApplicationContextAware;
import org.springframework.context.event.ContextRefreshedEvent;
import org.springframework.context.event.EventListener;
import org.springframework.stereotype.Component;

@Component
public class AuditService implements ApplicationContextAware {

	public static ApplicationContext applicationContext = null;

	public static void autowire(final Object classToAutowire) {
		AuditService.applicationContext.getAutowireCapableBeanFactory().autowireBean(classToAutowire);
	}

	private final Map<Method, Object> prePersistMethods = new HashMap<>();
	private final Map<Method, Object> postPersistMethods = new HashMap<>();
	private final Map<Method, Object> preUpdateMethods = new HashMap<>();
	private final Map<Method, Object> postUpdateMethods = new HashMap<>();
	private final Map<Method, Object> preRemoveMethods = new HashMap<>();
	private final Map<Method, Object> postRemoveMethods = new HashMap<>();

	private final Map<Method, Object> postLoadMethods = new HashMap<>();

	public ApplicationContext getApplicationContext() {
		return AuditService.applicationContext;
	}

	@EventListener
	public void handleContextRefresh(final ContextRefreshedEvent event) {
		final String[] beanNames = AuditService.applicationContext.getBeanDefinitionNames();
		for (final String beanName : beanNames) {
			final Object bean = AuditService.applicationContext.getBean(beanName);
			for (final Method method : bean.getClass().getMethods()) {
				if (method.isAnnotationPresent(PrePersist.class)) {
					this.prePersistMethods.put(method, bean);
				}
				if (method.isAnnotationPresent(PostPersist.class)) {
					this.postPersistMethods.put(method, bean);
				}
				if (method.isAnnotationPresent(PreRemove.class)) {
					this.preRemoveMethods.put(method, bean);
				}
				if (method.isAnnotationPresent(PostRemove.class)) {
					this.postRemoveMethods.put(method, bean);
				}
				if (method.isAnnotationPresent(PreUpdate.class)) {
					this.preUpdateMethods.put(method, bean);
				}
				if (method.isAnnotationPresent(PostUpdate.class)) {
					this.postUpdateMethods.put(method, bean);
				}
				if (method.isAnnotationPresent(PostLoad.class)) {
					this.postLoadMethods.put(method, bean);
				}
			}
		}
	}

	private void invokeMethod(final Method method, final Object bean, final Object object)
			throws IllegalAccessException, InvocationTargetException {
		final Class<?>[] typeParameters = method.getParameterTypes();
		if (typeParameters.length == 1 && typeParameters[0].isInstance(object)) {
			method.invoke(bean, object);
		}
	}

	public void postLoad(final Object object) throws IllegalAccessException, InvocationTargetException {
		for (final Method method : this.postLoadMethods.keySet()) {
			this.invokeMethod(method, this.postLoadMethods.get(method), object);
		}
	}

	public void postPersist(final Object object) throws IllegalAccessException, InvocationTargetException {
		for (final Method method : this.postPersistMethods.keySet()) {
			this.invokeMethod(method, this.postPersistMethods.get(method), object);
		}

	}

	public void postRemove(final Object object) throws IllegalAccessException, InvocationTargetException {
		for (final Method method : this.postRemoveMethods.keySet()) {
			this.invokeMethod(method, this.postRemoveMethods.get(method), object);
		}

	}

	public void postUpdate(final Object object) throws IllegalAccessException, InvocationTargetException {
		for (final Method method : this.postUpdateMethods.keySet()) {
			this.invokeMethod(method, this.postUpdateMethods.get(method), object);
		}

	}

	public void prePersist(final Object object) throws IllegalAccessException, InvocationTargetException {
		for (final Method method : this.prePersistMethods.keySet()) {
			this.invokeMethod(method, this.prePersistMethods.get(method), object);
		}

	}

	public void preRemove(final Object object) throws IllegalAccessException, InvocationTargetException {
		for (final Method method : this.preRemoveMethods.keySet()) {
			this.invokeMethod(method, this.preRemoveMethods.get(method), object);
		}
	}

	public void preUpdate(final Object object) throws IllegalAccessException, InvocationTargetException {
		for (final Method method : this.preUpdateMethods.keySet()) {
			this.invokeMethod(method, this.preUpdateMethods.get(method), object);
		}
	}

	@Override
	public void setApplicationContext(final ApplicationContext ac) throws BeansException {
		AuditService.applicationContext = ac;
	}

}
