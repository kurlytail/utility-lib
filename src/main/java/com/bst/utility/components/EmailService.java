package com.bst.utility.components;

import java.util.Locale;

import javax.mail.MessagingException;
import javax.mail.internet.MimeMessage;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.mail.MailSender;
import org.springframework.mail.javamail.JavaMailSender;
import org.springframework.mail.javamail.MimeMessageHelper;
import org.springframework.stereotype.Service;
import org.thymeleaf.context.Context;
import org.thymeleaf.spring5.SpringTemplateEngine;

@Service
public class EmailService {

	@Autowired
	private MailSender mailSender;
	
	@Autowired
	private SpringTemplateEngine templateEngine;

	public MailSender getMailSender() {
		return mailSender;
	}

	public void setMailSender(MailSender mailSender) {
		this.mailSender = mailSender;
	}

	public void sendMessage(final String[] recepients, final String template, 
			final String from, final String subject,
			final Object dto, final Locale locale)
			throws MessagingException {

		final Context ctx = new Context(locale);
		ctx.setVariable("dto", dto);
		
		final JavaMailSender javaMailSender = (JavaMailSender)getMailSender();
		final MimeMessage mimeMessage = javaMailSender.createMimeMessage();
		final MimeMessageHelper helper = new MimeMessageHelper(mimeMessage, false, "utf-8");
	    final String htmlContent = this.templateEngine.process(template, ctx);
	    
		helper.setText(htmlContent, true);
		helper.setSubject(subject);
		helper.setFrom(from);
		helper.setTo(recepients);
		
		javaMailSender.send(mimeMessage);
	}
}
