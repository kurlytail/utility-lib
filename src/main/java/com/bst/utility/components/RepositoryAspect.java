package com.bst.utility.components;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.HashMap;
import java.util.Map;

import javax.persistence.PostLoad;
import javax.persistence.PostPersist;
import javax.persistence.PostRemove;
import javax.persistence.PostUpdate;
import javax.persistence.PrePersist;
import javax.persistence.PreRemove;
import javax.persistence.PreUpdate;

import org.aspectj.lang.JoinPoint;
import org.aspectj.lang.annotation.After;
import org.aspectj.lang.annotation.Aspect;
import org.aspectj.lang.annotation.Before;
import org.aspectj.lang.annotation.Pointcut;
import org.springframework.context.event.ContextRefreshedEvent;
import org.springframework.context.event.EventListener;

@Aspect
public class RepositoryAspect {

	private final Map<Method, Object> postLoadMethods = new HashMap<>();

	private final Map<Method, Object> postPersistMethods = new HashMap<>();

	private final Map<Method, Object> postRemoveMethods = new HashMap<>();

	private final Map<Method, Object> postUpdateMethods = new HashMap<>();

	private final Map<Method, Object> prePersistMethods = new HashMap<>();

	private final Map<Method, Object> preRemoveMethods = new HashMap<>();
	private final Map<Method, Object> preUpdateMethods = new HashMap<>();

	@EventListener
	public void handleContextRefresh(final ContextRefreshedEvent event) {
		final String[] beanNames = event.getApplicationContext().getBeanDefinitionNames();
		for (final String beanName : beanNames) {
			final Object bean = event.getApplicationContext().getBean(beanName);
			for (final Method method : bean.getClass().getMethods()) {
				if (method.isAnnotationPresent(PrePersist.class)) {
					this.prePersistMethods.put(method, bean);
				}
				if (method.isAnnotationPresent(PostPersist.class)) {
					this.postPersistMethods.put(method, bean);
				}
				if (method.isAnnotationPresent(PreRemove.class)) {
					this.preRemoveMethods.put(method, bean);
				}
				if (method.isAnnotationPresent(PostRemove.class)) {
					this.postRemoveMethods.put(method, bean);
				}
				if (method.isAnnotationPresent(PreUpdate.class)) {
					this.preUpdateMethods.put(method, bean);
				}
				if (method.isAnnotationPresent(PostUpdate.class)) {
					this.postUpdateMethods.put(method, bean);
				}
				if (method.isAnnotationPresent(PostLoad.class)) {
					this.postLoadMethods.put(method, bean);
				}
			}
		}
	}

	private void invokeMethod(final Method method, final Object bean, final Object object)
			throws IllegalAccessException, InvocationTargetException {
		final Class<?>[] typeParameters = method.getParameterTypes();
		if ((typeParameters.length == 1) && typeParameters[0].isInstance(object)) {
			method.invoke(bean, object);
		}
	}

	@Pointcut("execution(* javax.persistence.EntityManager.merge(..)) && args(entity)")
	public void mergeMethod(final Object entity) {
	}

	@Pointcut("execution(* javax.persistence.EntityManager.persist(..)) && args(entity)")
	public void persistMethod(final Object entity) {
	}

	@After("persistMethod(entity)")
	public void postPersist(final JoinPoint jp, final Object entity) throws Throwable {
		for (final Method method : this.postPersistMethods.keySet()) {
			this.invokeMethod(method, this.postPersistMethods.get(method), entity);
		}
	}

	@After("removeMethod(entity)")
	public void postRemove(final JoinPoint jp, final Object entity) throws Throwable {
		for (final Method method : this.postRemoveMethods.keySet()) {
			this.invokeMethod(method, this.postRemoveMethods.get(method), entity);
		}
	}

	@After("mergeMethod(entity)")
	public void postUpdate(final JoinPoint jp, final Object entity) throws Throwable {
		for (final Method method : this.postUpdateMethods.keySet()) {
			this.invokeMethod(method, this.postUpdateMethods.get(method), entity);
		}
	}

	@Before("persistMethod(entity)")
	public void prePersist(final JoinPoint jp, final Object entity) throws Throwable {
		for (final Method method : this.prePersistMethods.keySet()) {
			this.invokeMethod(method, this.prePersistMethods.get(method), entity);
		}
	}

	@Before("removeMethod(entity)")
	public void preRemove(final JoinPoint jp, final Object entity) throws Throwable {
		for (final Method method : this.preRemoveMethods.keySet()) {
			this.invokeMethod(method, this.preRemoveMethods.get(method), entity);
		}
	}

	@Before("mergeMethod(entity)")
	public void preUpdate(final JoinPoint jp, final Object entity) throws Throwable {
		for (final Method method : this.preUpdateMethods.keySet()) {
			this.invokeMethod(method, this.preUpdateMethods.get(method), entity);
		}
	}

	@Pointcut("execution(* javax.persistence.EntityManager.remove(..)) && args(entity)")
	public void removeMethod(final Object entity) {
	}
}
