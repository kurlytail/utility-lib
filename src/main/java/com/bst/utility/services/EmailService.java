package com.bst.utility.services;

import java.net.InetAddress;
import java.net.MalformedURLException;
import java.net.URL;
import java.net.UnknownHostException;
import java.util.Locale;

import javax.mail.MessagingException;
import javax.mail.internet.MimeMessage;
import javax.servlet.ServletContext;
import javax.servlet.http.HttpServletRequest;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.web.servlet.context.ServletWebServerInitializedEvent;
import org.springframework.context.ApplicationListener;
import org.springframework.mail.MailSender;
import org.springframework.mail.javamail.JavaMailSender;
import org.springframework.mail.javamail.MimeMessageHelper;
import org.springframework.stereotype.Service;
import org.springframework.web.context.request.RequestAttributes;
import org.springframework.web.context.request.RequestContextHolder;
import org.springframework.web.context.request.ServletRequestAttributes;
import org.thymeleaf.context.Context;
import org.thymeleaf.spring5.SpringTemplateEngine;

@Service
public class EmailService implements ApplicationListener<ServletWebServerInitializedEvent> {

	@Autowired
	private MailSender mailSender;

	@Autowired
	private SpringTemplateEngine templateEngine;
	
	@Autowired
	private ServletContext servletContext;
	
	private String prefix;

	public MailSender getMailSender() {
		return mailSender;
	}

	public void setMailSender(MailSender mailSender) {
		this.mailSender = mailSender;
	}

	public void sendMessage(final String[] recepients, final String template, final String from, final String subject,
			final String dtoName, final Object dto, final Locale locale) throws MessagingException, MalformedURLException {
		
		checkPrefix();

		final JavaMailSender javaMailSender = (JavaMailSender) getMailSender();
		final MimeMessage mimeMessage = javaMailSender.createMimeMessage();
		final MimeMessageHelper helper = new MimeMessageHelper(mimeMessage, false, "utf-8");

		final Context context = new Context(locale);
		context.setVariable(dtoName, dto);
		context.setVariable("DomainURL", prefix);

		final String htmlContent = this.templateEngine.process(template, context);

		helper.setText(htmlContent, true);
		helper.setSubject(subject);
		helper.setFrom(from);
		helper.setTo(recepients);

		javaMailSender.send(mimeMessage);
	}

	private void checkPrefix() throws MalformedURLException {
		RequestAttributes requestAttributes = RequestContextHolder.getRequestAttributes();
	    if (requestAttributes instanceof ServletRequestAttributes) {
	    	HttpServletRequest request = ((ServletRequestAttributes)requestAttributes).getRequest();
	    	int port = request.getServerPort();

	    	if (request.getScheme().equals("http") && port == 80) {
	    	    port = -1;
	    	} else if (request.getScheme().equals("https") && port == 443) {
	    	    port = -1;
	    	}

	    	URL serverURL = new URL(request.getScheme(), request.getServerName(), port, servletContext.getContextPath());
	    	this.prefix = serverURL.toString();
	    }
	}

	@Override
	public void onApplicationEvent(ServletWebServerInitializedEvent event) {
		int port = event.getWebServer().getPort();
		try {
			String serverName = InetAddress.getLocalHost().getHostName();
			prefix = "http://" + serverName;
			if (port == 443) {
				prefix = "https://" + serverName;
			} else if (port != 22 && port != 0) {
				prefix = prefix + ":" + port;
			}
			prefix = prefix + servletContext.getContextPath();
		} catch (UnknownHostException e) {
			e.printStackTrace();
		}
	}
}
