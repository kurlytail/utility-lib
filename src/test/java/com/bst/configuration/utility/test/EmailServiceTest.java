package com.bst.configuration.utility.test;

import static io.github.jsonSnapshot.SnapshotMatcher.expect;
import static io.github.jsonSnapshot.SnapshotMatcher.start;
import static io.github.jsonSnapshot.SnapshotMatcher.validateSnapshots;

import java.io.IOException;
import java.util.Locale;

import javax.mail.MessagingException;
import javax.mail.internet.MimeMessage;

import org.junit.After;
import org.junit.AfterClass;
import org.junit.BeforeClass;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.autoconfigure.EnableAutoConfiguration;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.junit4.SpringJUnit4ClassRunner;

import com.bst.configuration.utility.UtilityConfiguration;
import com.bst.utility.components.EmailService;
import com.icegreen.greenmail.junit.GreenMailRule;
import com.icegreen.greenmail.util.ServerSetupTest;

@RunWith(SpringJUnit4ClassRunner.class)
@SpringBootTest(classes = UtilityConfiguration.class)
@EnableAutoConfiguration
public class EmailServiceTest {
	
	@Autowired
	private EmailService emailService;

	@Rule
	public GreenMailRule smtpServerRule = new GreenMailRule(ServerSetupTest.ALL);

	@After
	public void cleanup() {
		this.smtpServerRule.stop();
	}

	@BeforeClass
	public static void startSnapshot() {
		start();
	}

	@AfterClass
	public static void stopSnapshot() {
		validateSnapshots();
	}

	@Test
	public void sendAndReceiveEmailTest() throws IOException, MessagingException {
		
		emailService.sendMessage(new String[] { "some@email" }, "test-email.html", "my@email", "MySubject", "dto", new Object() {
			public String testField1 = "test1";
			public String testField2 = "test2";
		}, Locale.ENGLISH);

		MimeMessage[] messages = smtpServerRule.getReceivedMessages();

		assert (messages.length == 1);

		for (MimeMessage msg : messages) {
			expect(msg.getContent(), msg.getAllHeaders(), msg.getAllRecipients(), msg.getFrom()).toMatchSnapshot();
		}
	}
}
