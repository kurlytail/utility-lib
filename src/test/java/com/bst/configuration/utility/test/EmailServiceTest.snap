{
  "sendAndReceiveEmailTest" : [ "\"<!DOCTYPE html>\\r\\n<html>\\r\\n<body>\\r\\n    <div>test1</div>\\r\\n    <div>test2</div>\\r\\n</body>\\r\\n</html>\\r\\n\"", "[{\"address\":\"some@email\",\"personal\":null,\"type\":\"rfc822\",\"group\":false}]", "[{\"address\":\"my@email\",\"personal\":null,\"type\":\"rfc822\",\"group\":false}]" ]
}